### DISCUSSION TITLE
304 and 412 HTTP response statuses

### DISCUSSION THREAD

**Discussion by pboguslawski** [swaggest/rest#145] [General]
There is no status defined for HTTP 304 nor 412 in

https://github.com/swaggest/usecase/blob/master/status/status.go#L12 (*)

Any way to use `status.Wrap` to respond with 304 or 412 if `If-None-Match` precondition fails which is required by

https://www.rfc-editor.org/rfc/rfc7232#section-3.2

Maybe additional states (substates?) should be added to (*)?

**Update**:
304 should be treated as kind of "success" response, [without body](https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.3.5), so probably should not be registred with SetExpectedErrors.

Any way to specify multiple possible success responses for given interactor (200 or 204 or 304 for example)?

When using custom error stuct (wraps original error with http status code + context fields) and custom error responses it is seems possible to workaround swaggest status limitiations with something like
```
u.SetExpectedErrors(
		errors.Wrap(nil, http.StatusBadRequest, nil),
		errors.Wrap(nil, http.StatusPreconditionFailed, nil),
		errors.Wrap(nil, http.StatusInternalServerError, nil),
		errors.Wrap(nil, http.StatusNotModified, nil),
	)
```
but this generates 304 response in openapi.json with default error response body which is wrong (there is no body in 304 responses). When defining empty response for 304 in custom MakeErrResp
```
	if code == http.StatusNotModified {
		return code, nil // or struct{}{}
	}
```
final openapi.json still contains body definition and default error response content type (here `application/problem+json`) for 304
```
"304": {
      "description": "Not Modified",
      "content": {
       "application/problem+json": {
        "schema": {
         "type": "string"
        }
       }
      }
     },
```
There should not be body in openapi.json problaby when `MakeErrResp` gets `nil`; and maybe `MakeErrResp` should have separate param for defining response content type to allow one to define different error responses with different content types (no content type if `nil` like for 304).

**Comment by vearutop** [comment]
New version improves support for this case, please check updated example at https://github.com/swaggest/rest/blob/v0.2.44/_examples/advanced-generic/output_writer.go.

**Reply by pboguslawski** [reply]
After #147 and #148 it's possible to have 304 (empty body as required) and 412 responses. Thank you!
